File: src/mycli/services/hooks/runner.py
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
import os
from pathlib import Path
import subprocess
from typing import Any, Callable

from mycli.services.hooks.allowlist import HookAllowlistStatus
from mycli.services.hooks.config import (
    ConfiguredHookSpec,
    HookEnvPolicy,
    HookWorkingDirectory,
)
from mycli.services.hooks.types import HookAction, HookContext, HookResult
from mycli.services.hooks.types import HookPoint

MAX_HOOK_OUTPUT_CHARS = 2000
MAX_HOOK_MESSAGE_CHARS = 200
MAX_MODIFIED_ARGS = 20

_SAFE_ENV_KEYS = ("HOME", "PATH", "SHELL", "TMPDIR", "USER")
_MINIMAL_PATH = "/usr/bin:/bin:/usr/sbin:/sbin"


@dataclass(slots=True, frozen=True)
class ConfiguredHookRunSummary:
    execution_id: str
    hook_id: str
    hook_name: str
    hook_point: str
    status: str
    action: str | None
    duration_ms: int
    exit_code: int | None = None
    stdout_chars: int = 0
    stderr_chars: int = 0
    message: str = ""

    def safe_payload(self) -> dict[str, object]:
        payload: dict[str, object] = {
            "execution_id": self.execution_id,
            "hook_id": self.hook_id,
            "hook_name": self.hook_name,
            "hook_point": self.hook_point,
            "status": self.status,
            "duration_ms": self.duration_ms,
            "stdout_chars": self.stdout_chars,
            "stderr_chars": self.stderr_chars,
        }
        if self.action is not None:
            payload["action"] = self.action
        if self.exit_code is not None:
            payload["exit_code"] = self.exit_code
        if self.message:
            payload["message"] = _safe_message(self.message)
        return payload


@dataclass(slots=True)
class ConfiguredHookCallback:
    spec: ConfiguredHookSpec
    workspace_root: Path
    monotonic: Any
    trace_sink: Any | None = None
    allowlist_status: Callable[[ConfiguredHookSpec], HookAllowlistStatus] | None = None

    def __call__(self, ctx: HookContext) -> HookResult:
        source = ctx.metadata.get("source")
        if not self.spec.enabled or not self.spec.matches(
            tool_name=ctx.tool_name,
            source=str(source) if isinstance(source, str) else None,
        ):
            return HookResult(action=HookAction.ALLOW)
        started_at = float(self.monotonic())
        allowlist_status = (
            self.allowlist_status(self.spec)
            if self.allowlist_status is not None
            else HookAllowlistStatus(allowed=True, reason="not_configured", digest="")
        )
        if not allowlist_status.allowed:
            summary = ConfiguredHookRunSummary(
                execution_id=_execution_id(self.spec, ctx),
                hook_id=self.spec.hook_id,
                hook_name=self.spec.name,
                hook_point=self.spec.hook_point.value,
                status="error",
                action=HookAction.ERROR.value,
                duration_ms=_duration_ms(self.monotonic, started_at),
                message=f"not allowlisted: {allowlist_status.reason}",
            )
            _emit_trace(self.trace_sink, ctx, summary)
            return HookResult(action=HookAction.ERROR, message="configured hook not allowlisted")
        try:
            completed = subprocess.run(
                list(self.spec.command),
                input=json.dumps(_hook_payload(ctx), ensure_ascii=False),
                text=True,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                timeout=self.spec.timeout_seconds,
                cwd=str(_working_directory(self.spec, self.workspace_root)),
                env=_hook_env(self.spec),
                check=False,
            )
        except subprocess.TimeoutExpired:
            duration_ms = _duration_ms(self.monotonic, started_at)
            summary = ConfiguredHookRunSummary(
                execution_id=_execution_id(self.spec, ctx),
                hook_id=self.spec.hook_id,
                hook_name=self.spec.name,
                hook_point=self.spec.hook_point.value,
                status="error",
                action=HookAction.ERROR.value,
                duration_ms=duration_ms,
                message="timeout",
            )
            _emit_trace(self.trace_sink, ctx, summary)
            return HookResult(action=HookAction.ERROR, message="configured hook timed out")
        except OSError as exc:
            duration_ms = _duration_ms(self.monotonic, started_at)
            summary = ConfiguredHookRunSummary(
                execution_id=_execution_id(self.spec, ctx),
                hook_id=self.spec.hook_id,
                hook_name=self.spec.name,
                hook_point=self.spec.hook_point.value,
                status="error",
                action=HookAction.ERROR.value,
                duration_ms=duration_ms,
                message=exc.__class__.__name__,
            )
            _emit_trace(self.trace_sink, ctx, summary)
            return HookResult(action=HookAction.ERROR, message="configured hook failed")

        duration_ms = _duration_ms(self.monotonic, started_at)
        stdout = _bounded_text(completed.stdout)
        stderr = _bounded_text(completed.stderr)
        result, summary = _result_from_completed_process(
            spec=self.spec,
            execution_id=_execution_id(self.spec, ctx),
            stdout=stdout,
            stderr=stderr,
            exit_code=completed.returncode,
            duration_ms=duration_ms,
        )
        _emit_trace(self.trace_sink, ctx, summary)
        return result


def _result_from_completed_process(
    *,
    spec: ConfiguredHookSpec,
    execution_id: str,
    stdout: str,
    stderr: str,
    exit_code: int,
    duration_ms: int,
) -> tuple[HookResult, ConfiguredHookRunSummary]:
    if exit_code == 2 and spec.hook_point in _BLOCKING_EXIT_CODE_HOOK_POINTS:
        message = _safe_message(stderr or stdout or "blocked by configured hook")
        summary = ConfiguredHookRunSummary(
            execution_id=execution_id,
            hook_id=spec.hook_id,
            hook_name=spec.name,
            hook_point=spec.hook_point.value,
            status="ok",
            action=HookAction.DENY.value,
            duration_ms=duration_ms,
            exit_code=exit_code,
            stdout_chars=len(stdout),
            stderr_chars=len(stderr),
            message=message,
        )
        return HookResult(action=HookAction.DENY, message=message), summary
    if exit_code != 0:
        summary = ConfiguredHookRunSummary(
            execution_id=execution_id,
            hook_id=spec.hook_id,
            hook_name=spec.name,
            hook_point=spec.hook_point.value,
            status="error",
            action=HookAction.ERROR.value,
            duration_ms=duration_ms,
            exit_code=exit_code,
            stdout_chars=len(stdout),
            stderr_chars=len(stderr),
            message=stderr or stdout or f"exit={exit_code}",
        )
        return HookResult(action=HookAction.ERROR, message="configured hook failed"), summary
    stripped_stdout = stdout.strip()
    try:
        payload = json.loads(stripped_stdout) if stripped_stdout else {}
    except json.JSONDecodeError:
        if (
            spec.hook_point in _PLAIN_TEXT_ADDITIONAL_CONTEXT_HOOK_POINTS
            and not _looks_like_json(stripped_stdout)
        ):
            context = _safe_context(stripped_stdout)
            summary = ConfiguredHookRunSummary(
                execution_id=execution_id,
                hook_id=spec.hook_id,
                hook_name=spec.name,
                hook_point=spec.hook_point.value,
                status="ok",
                action=HookAction.ALLOW.value,
                duration_ms=duration_ms,
                exit_code=exit_code,
                stdout_chars=len(stdout),
                stderr_chars=len(stderr),
            )
            return (
                HookResult(
                    action=HookAction.ALLOW,
                    additional_contexts=(context,) if context else (),
                ),
                summary,
            )
        summary = ConfiguredHookRunSummary(
            execution_id=execution_id,
            hook_id=spec.hook_id,
            hook_name=spec.name,
            hook_point=spec.hook_point.value,
            status="error",
            action=HookAction.ERROR.value,
            duration_ms=duration_ms,
            exit_code=exit_code,
            stdout_chars=len(stdout),
            stderr_chars=len(stderr),
            message="stdout was not JSON",
        )
        return HookResult(action=HookAction.ERROR, message="configured hook output invalid"), summary
    if not isinstance(payload, dict):
        return _invalid_action_result(
            spec=spec,
            execution_id=execution_id,
            duration_ms=duration_ms,
            stdout=stdout,
            stderr=stderr,
        )
    hook_specific = payload.get("hookSpecificOutput")
    codex_permission_deny = (
        isinstance(hook_specific, dict)
        and hook_specific.get("permissionDecision") == "deny"
    )
    codex_block = (
        payload.get("decision") == "block"
        or payload.get("continue") is False
        or codex_permission_deny
    )
    if codex_block:
        action = HookAction.DENY
    else:
        raw_action = payload.get("action", HookAction.ALLOW.value)
        try:
            action = HookAction(str(raw_action))
        except ValueError:
            return _invalid_action_result(
                spec=spec,
                execution_id=execution_id,
                duration_ms=duration_ms,
                stdout=stdout,
                stderr=stderr,
            )
    message = str(
        payload.get("message")
        or payload.get("reason")
        or payload.get("stopReason")
        or (
            hook_specific.get("permissionDecisionReason")
            if isinstance(hook_specific, dict)
            else None
        )
        or ""
    )
    modified_args = _modified_args(payload.get("modified_args")) if action is HookAction.MODIFY else None
    additional_contexts = _additional_contexts(payload.get("additional_contexts"))
    additional_contexts = (
        *additional_contexts,
        *_codex_additional_contexts(hook_specific, hook_point=spec.hook_point),
    )
    summary = ConfiguredHookRunSummary(
        execution_id=execution_id,
        hook_id=spec.hook_id,
        hook_name=spec.name,
        hook_point=spec.hook_point.value,
        status="ok",
        action=action.value,
        duration_ms=duration_ms,
        exit_code=exit_code,
        stdout_chars=len(stdout),
        stderr_chars=len(stderr),
        message=message,
    )
    return (
        HookResult(
            action=action,
            message=_safe_message(message),
            modified_args=modified_args,
            additional_contexts=additional_contexts,
        ),
        summary,
    )


def _invalid_action_result(
    *,
    spec: ConfiguredHookSpec,
    execution_id: str,
    duration_ms: int,
    stdout: str,
    stderr: str,
) -> tuple[HookResult, ConfiguredHookRunSummary]:
    summary = ConfiguredHookRunSummary(
        execution_id=execution_id,
        hook_id=spec.hook_id,
        hook_name=spec.name,
        hook_point=spec.hook_point.value,
        status="error",
        action=HookAction.ERROR.value,
        duration_ms=duration_ms,
        exit_code=0,
        stdout_chars=len(stdout),
        stderr_chars=len(stderr),
        message="unsupported hook action",
    )
    return HookResult(action=HookAction.ERROR, message="configured hook output invalid"), summary


def _modified_args(value: object) -> dict[str, Any] | None:
    if not isinstance(value, dict):
        return None
    modified: dict[str, Any] = {}
    for key, item in value.items():
        if not isinstance(key, str):
            continue
        if len(modified) >= MAX_MODIFIED_ARGS:
            break
        modified[key] = item
    return modified or None


def _additional_contexts(value: object) -> tuple[str, ...]:
    if isinstance(value, str):
        normalized = _safe_context(value)
        return (normalized,) if normalized else ()
    if not isinstance(value, list):
        return ()
    contexts: list[str] = []
    for item in value:
        if not isinstance(item, str):
            continue
        normalized = _safe_context(item)
        if normalized:
            contexts.append(normalized)
    return tuple(contexts)


_BLOCKING_EXIT_CODE_HOOK_POINTS = frozenset(
    {
        HookPoint.PRE_TOOL_USE,
        HookPoint.POST_TOOL_USE,
        HookPoint.USER_PROMPT_SUBMIT,
        HookPoint.STOP,
    }
)
_PLAIN_TEXT_ADDITIONAL_CONTEXT_HOOK_POINTS = frozenset(
    {
        HookPoint.SESSION_START,
        HookPoint.USER_PROMPT_SUBMIT,
    }
)


def _codex_additional_contexts(value: object, *, hook_point: HookPoint) -> tuple[str, ...]:
    if hook_point not in _ADDITIONAL_CONTEXT_HOOK_POINTS:
        return ()
    if not isinstance(value, dict):
        return ()
    return _additional_contexts(value.get("additionalContext"))


_ADDITIONAL_CONTEXT_HOOK_POINTS = frozenset(
    {
        HookPoint.POST_TOOL_USE,
        HookPoint.SESSION_START,
        HookPoint.USER_PROMPT_SUBMIT,
    }
)


def _hook_payload(ctx: HookContext) -> dict[str, object]:
    metadata_keys = sorted(str(key) for key in ctx.metadata)
    payload: dict[str, object] = {
        "hook_event_name": _codex_event_name(ctx.hook_point),
        "hook_point": ctx.hook_point.value,
        "tool_name": ctx.tool_name,
        "session_id": ctx.session_id,
        "metadata_keys": metadata_keys,
    }
    if ctx.tool_name is not None:
        payload["tool_input"] = dict(ctx.tool_args or {})
    prompt = ctx.metadata.get("prompt")
    if isinstance(prompt, str):
        payload["prompt"] = prompt
    source = ctx.metadata.get("source")
    if isinstance(source, str):
        payload["source"] = source
    return payload


def _codex_event_name(hook_point: HookPoint) -> str:
    return {
        HookPoint.PRE_TOOL_USE: "PreToolUse",
        HookPoint.POST_TOOL_USE: "PostToolUse",
        HookPoint.SESSION_START: "SessionStart",
        HookPoint.USER_PROMPT_SUBMIT: "UserPromptSubmit",
        HookPoint.STOP: "Stop",
    }.get(hook_point, hook_point.value)


def _execution_id(spec: ConfiguredHookSpec, ctx: HookContext) -> str:
    parts = (
        spec.source.value,
        spec.hook_id,
        spec.hook_point.value,
        str(ctx.session_id or ""),
        str(ctx.metadata.get("turn_id") or ""),
        str(ctx.tool_name or ""),
    )
    digest = hashlib.sha256("\0".join(parts).encode("utf-8")).hexdigest()[:16]
    return f"hookexec_{digest}"


def _hook_env(spec: ConfiguredHookSpec) -> dict[str, str]:
    env = {
        "PATH": os.environ.get("PATH") or _MINIMAL_PATH,
        "MYCLI_HOOK_ID": spec.hook_id,
        "MYCLI_HOOK_POINT": spec.hook_point.value,
        "MYCLI_HOOK_SOURCE": spec.source.value,
    }
    if spec.env_policy is HookEnvPolicy.INHERIT_SAFE:
        for key in _SAFE_ENV_KEYS:
            value = os.environ.get(key)
            if value:
                env[key] = value
    return env


def _working_directory(spec: ConfiguredHookSpec, workspace_root: Path) -> Path:
    if spec.working_directory is HookWorkingDirectory.CONFIG and spec.source_path is not None:
        return spec.source_path.parent
    return workspace_root


def _duration_ms(monotonic: Any, started_at: float) -> int:
    return max(0, int(round((float(monotonic()) - started_at) * 1000)))


def _bounded_text(value: str | None) -> str:
    if not value:
        return ""
    return value[:MAX_HOOK_OUTPUT_CHARS]


def _safe_message(message: str) -> str:
    normalized = " ".join(message.split())
    lowered = normalized.lower()
    if any(part in lowered for part in ("api_key", "apikey", "token", "secret", "password", "bearer ")):
        return "redacted"
    return normalized[:MAX_HOOK_MESSAGE_CHARS]


def _safe_context(message: str) -> str:
    normalized = message.strip()
    lowered = normalized.lower()
    if any(part in lowered for part in ("api_key", "apikey", "token", "secret", "password", "bearer ")):
        return "redacted"
    return normalized[:MAX_HOOK_OUTPUT_CHARS]


def _looks_like_json(value: str) -> bool:
    stripped = value.lstrip()
    return stripped.startswith("{") or stripped.startswith("[")


def _emit_trace(trace_sink: Any | None, ctx: HookContext, summary: ConfiguredHookRunSummary) -> None:
    if trace_sink is None:
        return
    trace_sink(ctx, summary)

File: backend/packages/integrations/src/hooks/runner.ts
import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { dirname } from "node:path";
import type { HookInvocation, HookResult } from "@mycli/core";
import {
	createProcessController,
	prepareSandboxedProcess,
	type SandboxedProcessLaunch,
	type SandboxProfile,
} from "@mycli/tools";
import { configuredHookMatches } from "./config.ts";
import type { HookAllowlistStore } from "./allowlist.ts";
import type {
	ConfiguredHookExecutorContract,
	ConfiguredHookSpec,
	ConfiguredHookTraceSummary,
} from "./types.ts";

export interface ConfiguredHookRunnerOptions {
	readonly workspaceRoot: string;
	readonly allowlistStore: Pick<HookAllowlistStore, "statusFor">;
	readonly sandboxProfile: (cwd: string) => SandboxProfile;
	readonly env?: Readonly<NodeJS.ProcessEnv>;
	readonly onTrace?: (trace: ConfiguredHookTraceSummary) => void;
	readonly monotonic?: () => number;
	readonly prepareProcess?: (
		argv: readonly string[],
		profile: SandboxProfile,
	) => SandboxedProcessLaunch;
}

interface ProcessOutcome {
	readonly kind: "exit" | "timeout" | "overflow" | "spawn_error" | "interrupted";
	readonly exitCode?: number;
	readonly stdout: CaptureSnapshot;
	readonly stderr: CaptureSnapshot;
}

interface CaptureSnapshot {
	readonly text: string;
	readonly chars: number;
	readonly truncated: boolean;
}

const MAX_INPUT_CHARS = 65_536;
const MAX_OUTPUT_CHARS = 2_000;
const MAX_CONTEXTS = 8;
const MAX_MODIFIED_ARGUMENTS = 20;
const MINIMAL_PATH = process.platform === "win32"
	? "C:\\Windows\\System32;C:\\Windows"
	: "/usr/bin:/bin:/usr/sbin:/sbin";
const SAFE_ENV_KEYS = ["HOME", "PATH", "SHELL", "TMPDIR", "USER"] as const;
const SENSITIVE_TEXT = /\b(?:Bearer\s+\S+|sk-[A-Za-z0-9_-]{8,}|(?:api[_-]?key|authorization|cookie|password|secret|token)\s*[:=]\s*\S+)/iu;
const BLOCKING_EXIT_POINTS = new Set([
	"pre_tool_use",
	"post_tool_use",
	"user_prompt_submit",
	"stop",
]);
const PLAIN_CONTEXT_POINTS = new Set(["session_start", "user_prompt_submit"]);
const CODEX_CONTEXT_POINTS = new Set(["post_tool_use", "session_start", "user_prompt_submit"]);

export class ConfiguredHookRunner implements ConfiguredHookExecutorContract {
	readonly #options: ConfiguredHookRunnerOptions;
	readonly #env: Readonly<NodeJS.ProcessEnv>;
	readonly #monotonic: () => number;
	readonly #prepareProcess: NonNullable<ConfiguredHookRunnerOptions["prepareProcess"]>;

	constructor(options: ConfiguredHookRunnerOptions) {
		if (!options.workspaceRoot.trim()) throw new TypeError("workspaceRoot must be non-empty");
		this.#options = options;
		this.#env = options.env ?? process.env;
		this.#monotonic = options.monotonic ?? (() => performance.now());
		this.#prepareProcess = options.prepareProcess ?? prepareSandboxedProcess;
	}

	async run(
		spec: ConfiguredHookSpec,
		invocation: HookInvocation,
		signal: AbortSignal,
	): Promise<HookResult> {
		assertNotAborted(signal);
		const startedAt = this.#monotonic();
		if (!spec.enabled || spec.hookPoint !== invocation.point || !configuredHookMatches(spec, {
			...(invocation.toolName ? { toolName: invocation.toolName } : {}),
			...(typeof invocation.metadata.source === "string"
				? { source: invocation.metadata.source }
				: {}),
		})) {
			const result = allowResult();
			this.#trace(spec, invocation, startedAt, result, emptyCapture(), emptyCapture(), "skipped");
			return result;
		}

		const approval = await this.#options.allowlistStore.statusFor(spec);
		assertNotAborted(signal);
		if (!approval.allowed) {
			const result = errorResult("configured hook not allowlisted");
			this.#trace(
				spec,
				invocation,
				startedAt,
				result,
				emptyCapture(),
				emptyCapture(),
				`not allowlisted: ${approval.reason}`,
			);
			return result;
		}

		let input: string;
		try {
			input = JSON.stringify(hookPayload(invocation));
		} catch {
			const result = errorResult("configured hook input invalid");
			this.#trace(spec, invocation, startedAt, result, emptyCapture(), emptyCapture(), "input_invalid");
			return result;
		}
		if (input.length > MAX_INPUT_CHARS) {
			const result = errorResult("configured hook input too large");
			this.#trace(spec, invocation, startedAt, result, emptyCapture(), emptyCapture(), "input_too_large");
			return result;
		}

		const cwd = spec.workingDirectory === "config" ? dirname(spec.configPath) : this.#options.workspaceRoot;
		let launch: SandboxedProcessLaunch;
		try {
			launch = this.#prepareProcess(spec.command, this.#options.sandboxProfile(cwd));
		} catch {
			const result = errorResult("configured hook failed");
			this.#trace(spec, invocation, startedAt, result, emptyCapture(), emptyCapture(), "sandbox_unavailable");
			return result;
		}

		const outcome = await runChildProcess({
			launch,
			cwd,
			env: hookEnvironment(spec, this.#env),
			input,
			timeoutMs: spec.timeoutMs,
			signal,
		});
		if (outcome.kind === "interrupted") {
			const result = errorResult("configured hook interrupted");
			this.#trace(
				spec,
				invocation,
				startedAt,
				result,
				outcome.stdout,
				outcome.stderr,
				"interrupted",
				undefined,
				"interrupted",
			);
			throw abortError();
		}

		const result = resultFromOutcome(spec, outcome);
		this.#trace(
			spec,
			invocation,
			startedAt,
			result.result,
			outcome.stdout,
			outcome.stderr,
			result.traceMessage,
			outcome.exitCode,
		);
		return result.result;
	}

	#trace(
		spec: ConfiguredHookSpec,
		invocation: HookInvocation,
		startedAt: number,
		result: HookResult,
		stdout: CaptureSnapshot,
		stderr: CaptureSnapshot,
		message: string,
		exitCode?: number,
		status?: ConfiguredHookTraceSummary["status"],
	): void {
		const trace = Object.freeze({
			executionId: executionId(spec, invocation),
			hookId: spec.hookId,
			hookName: spec.name,
			hookPoint: spec.hookPoint,
			status: status ?? (result.action === "error" ? "error" : "ok"),
			action: result.action,
			durationMs: Math.max(0, Math.round(this.#monotonic() - startedAt)),
			...(exitCode === undefined ? {} : { exitCode }),
			stdoutChars: stdout.chars,
			stderrChars: stderr.chars,
			stdoutTruncated: stdout.truncated,
			stderrTruncated: stderr.truncated,
			...(message ? { message: safeTraceMessage(message) } : {}),
		}) satisfies ConfiguredHookTraceSummary;
		try {
			this.#options.onTrace?.(trace);
		} catch {
			// Diagnostics must not affect hook behavior.
		}
	}
}

async function runChildProcess(options: {
	readonly launch: SandboxedProcessLaunch;
	readonly cwd: string;
	readonly env: Readonly<NodeJS.ProcessEnv>;
	readonly input: string;
	readonly timeoutMs: number;
	readonly signal: AbortSignal;
}): Promise<ProcessOutcome> {
	const child = spawn(options.launch.executable, [...options.launch.args], {
		cwd: options.cwd,
		env: { ...options.env },
		shell: false,
		windowsHide: true,
		detached: process.platform !== "win32",
		stdio: ["pipe", "pipe", "pipe"],
	});
	const stdout = new BoundedCapture(MAX_OUTPUT_CHARS);
	const stderr = new BoundedCapture(MAX_OUTPUT_CHARS);
	child.stdout.setEncoding("utf8");
	child.stderr.setEncoding("utf8");
	child.stdin.on("error", () => undefined);

	return new Promise((resolveOutcome) => {
		let finalizing = false;
		const controller = child.pid === undefined
			? undefined
			: createProcessController({
				pid: child.pid,
				exitCode: child.exitCode,
				signalCode: child.signalCode,
				kill: (signal) => child.kill(signal),
			});

		const finish = async (
			kind: ProcessOutcome["kind"],
			exitCode?: number,
			cleanup?: "interrupt" | "terminate",
		): Promise<void> => {
			if (finalizing) return;
			finalizing = true;
			clearTimeout(timer);
			options.signal.removeEventListener("abort", onAbort);
			if (cleanup && controller) {
				try {
					await controller[cleanup]();
				} catch {
					// The outcome remains fail-closed even if cleanup evidence is inconclusive.
				}
			}
			resolveOutcome(Object.freeze({
				kind,
				...(exitCode === undefined ? {} : { exitCode }),
				stdout: stdout.snapshot(),
				stderr: stderr.snapshot(),
			}));
		};
		const onAbort = (): void => { void finish("interrupted", undefined, "interrupt"); };
		const onOutput = (capture: BoundedCapture, chunk: string): void => {
			if (capture.append(chunk)) void finish("overflow", undefined, "terminate");
		};
		const timer = setTimeout(
			() => { void finish("timeout", undefined, "terminate"); },
			options.timeoutMs,
		);

		child.stdout.on("data", (chunk: string) => onOutput(stdout, chunk));
		child.stderr.on("data", (chunk: string) => onOutput(stderr, chunk));
		child.once("error", () => { void finish("spawn_error", undefined, "terminate"); });
		child.once("close", (code) => { void finish("exit", code ?? 1); });
		options.signal.addEventListener("abort", onAbort, { once: true });
		if (options.signal.aborted) {
			onAbort();
			return;
		}
		child.stdin.end(options.input, "utf8");
	});
}

class BoundedCapture {
	#text = "";
	#chars = 0;
	#truncated = false;

	constructor(readonly limit: number) {}

	append(chunk: string): boolean {
		const remaining = Math.max(0, this.limit - this.#text.length);
		if (remaining > 0) this.#text += chunk.slice(0, remaining);
		this.#chars = Math.min(this.limit + 1, this.#chars + chunk.length);
		if (chunk.length > remaining || this.#chars > this.limit) this.#truncated = true;
		return this.#truncated;
	}

	snapshot(): CaptureSnapshot {
		return Object.freeze({
			text: this.#text,
			chars: this.#chars,
			truncated: this.#truncated,
		});
	}
}

function resultFromOutcome(
	spec: ConfiguredHookSpec,
	outcome: ProcessOutcome,
): { readonly result: HookResult; readonly traceMessage: string } {
	if (outcome.kind === "timeout") {
		return { result: errorResult("configured hook timed out"), traceMessage: "timeout" };
	}
	if (outcome.kind === "overflow") {
		return { result: errorResult("configured hook output too large"), traceMessage: "output_too_large" };
	}
	if (outcome.kind === "spawn_error") {
		return { result: errorResult("configured hook failed"), traceMessage: "spawn_failed" };
	}
	const exitCode = outcome.exitCode ?? 1;
	if (exitCode === 2 && BLOCKING_EXIT_POINTS.has(spec.hookPoint)) {
		const message = safeMessage(
			outcome.stderr.text.trim() || outcome.stdout.text.trim() || "blocked by configured hook",
		);
		return { result: denyResult(message), traceMessage: message };
	}
	if (exitCode !== 0) {
		return { result: errorResult("configured hook failed"), traceMessage: "nonzero_exit" };
	}
	return resultFromOutput(spec, outcome.stdout.text.trim());
}

function resultFromOutput(
	spec: ConfiguredHookSpec,
	stdout: string,
): { readonly result: HookResult; readonly traceMessage: string } {
	let payload: unknown;
	try {
		payload = stdout ? JSON.parse(stdout) as unknown : {};
	} catch {
		if (PLAIN_CONTEXT_POINTS.has(spec.hookPoint) && !looksLikeJson(stdout)) {
			const context = safeContext(stdout);
			return {
				result: context ? allowResult([context]) : allowResult(),
				traceMessage: "",
			};
		}
		return { result: errorResult("configured hook output invalid"), traceMessage: "invalid_json" };
	}
	if (!isRecord(payload)) {
		return { result: errorResult("configured hook output invalid"), traceMessage: "invalid_result" };
	}
	const hookSpecific = isRecord(payload.hookSpecificOutput)
		? payload.hookSpecificOutput
		: undefined;
	const codexDenied = payload.decision === "block"
		|| payload.continue === false
		|| hookSpecific?.permissionDecision === "deny";
	const action = codexDenied ? "deny" : (payload.action ?? "allow");
	if (action !== "allow" && action !== "deny" && action !== "modify" && action !== "error") {
		return { result: errorResult("configured hook output invalid"), traceMessage: "invalid_action" };
	}
	const message = safeMessage(firstString(
		payload.message,
		payload.reason,
		payload.stopReason,
		hookSpecific?.permissionDecisionReason,
	) ?? "");
	if (action === "deny") {
		const deniedMessage = message || "blocked by configured hook";
		return { result: denyResult(deniedMessage), traceMessage: deniedMessage };
	}
	if (action === "error") {
		const errorMessage = message || "configured hook failed";
		return { result: errorResult(errorMessage), traceMessage: errorMessage };
	}
	if (action === "modify") {
		const argumentsValue = modifiedArguments(payload.modified_args ?? payload.arguments);
		if (!argumentsValue) {
			return { result: errorResult("configured hook output invalid"), traceMessage: "invalid_arguments" };
		}
		return {
			result: Object.freeze({ action: "modify", arguments: argumentsValue }),
			traceMessage: message,
		};
	}
	const contexts = additionalContexts(payload.additional_contexts, (
		CODEX_CONTEXT_POINTS.has(spec.hookPoint) ? hookSpecific?.additionalContext : undefined
	));
	return { result: allowResult(contexts), traceMessage: message };
}

function hookPayload(invocation: HookInvocation): Readonly<Record<string, unknown>> {
	const payload: Record<string, unknown> = {
		version: 1,
		hook_event_name: codexEventName(invocation.point),
		hook_point: invocation.point,
		tool_name: invocation.toolName ?? null,
		session_id: invocation.sessionId,
		turn_id: invocation.turnId,
		metadata_keys: Object.keys(invocation.metadata).sort(),
	};
	if (invocation.toolName) payload.tool_input = { ...invocation.arguments };
	if (typeof invocation.metadata.prompt === "string") payload.prompt = invocation.metadata.prompt;
	if (typeof invocation.metadata.source === "string") payload.source = invocation.metadata.source;
	return Object.freeze(payload);
}

function hookEnvironment(
	spec: ConfiguredHookSpec,
	environment: Readonly<NodeJS.ProcessEnv>,
): Readonly<NodeJS.ProcessEnv> {
	const env: NodeJS.ProcessEnv = {
		PATH: environmentValue(environment, "PATH") || MINIMAL_PATH,
		MYCLI_HOOK_ID: spec.hookId,
		MYCLI_HOOK_POINT: spec.hookPoint,
		MYCLI_HOOK_SOURCE: spec.scope,
	};
	if (spec.envPolicy === "inherit_safe") {
		for (const key of SAFE_ENV_KEYS) {
			const value = environmentValue(environment, key);
			if (value) env[key] = value;
		}
	}
	return Object.freeze(env);
}

function environmentValue(
	environment: Readonly<NodeJS.ProcessEnv>,
	key: string,
): string | undefined {
	const normalized = key.toLowerCase();
	for (const [name, value] of Object.entries(environment)) {
		if (name.toLowerCase() === normalized) return value;
	}
	return undefined;
}

function modifiedArguments(value: unknown): Readonly<Record<string, unknown>> | undefined {
	if (!isRecord(value)) return undefined;
	const entries = Object.entries(value).slice(0, MAX_MODIFIED_ARGUMENTS);
	return Object.freeze(Object.fromEntries(entries));
}

function additionalContexts(...values: readonly unknown[]): readonly string[] {
	const contexts: string[] = [];
	let remaining = MAX_OUTPUT_CHARS;
	for (const value of values) {
		const candidates = typeof value === "string" ? [value] : Array.isArray(value) ? value : [];
		for (const candidate of candidates) {
			if (typeof candidate !== "string" || contexts.length >= MAX_CONTEXTS || remaining <= 0) continue;
			const context = safeContext(candidate).slice(0, remaining);
			if (!context) continue;
			contexts.push(context);
			remaining -= context.length;
		}
	}
	return Object.freeze(contexts);
}

function allowResult(contexts: readonly string[] = []): HookResult {
	return contexts.length > 0
		? Object.freeze({ action: "allow", additionalContexts: Object.freeze([...contexts]) })
		: Object.freeze({ action: "allow" });
}

function denyResult(message: string): HookResult {
	return Object.freeze({ action: "deny", message });
}

function errorResult(message: string): HookResult {
	return Object.freeze({ action: "error", message });
}

function emptyCapture(): CaptureSnapshot {
	return Object.freeze({ text: "", chars: 0, truncated: false });
}

function safeMessage(value: string): string {
	const normalized = value.replace(/\s+/gu, " ").trim();
	return SENSITIVE_TEXT.test(normalized) ? "redacted" : normalized.slice(0, 200);
}

function safeTraceMessage(value: string): string {
	return safeMessage(value)
		.replace(/\b[A-Za-z]:\\(?:[^\\\s]+\\)*[^\\\s]*/gu, "[PATH]")
		.replace(/(^|[\s"'(])\/[^\s"'(),;]*/gu, "$1[PATH]");
}

function safeContext(value: string): string {
	const normalized = value.trim();
	return SENSITIVE_TEXT.test(normalized) ? "redacted" : normalized.slice(0, MAX_OUTPUT_CHARS);
}

function firstString(...values: readonly unknown[]): string | undefined {
	return values.find((value): value is string => typeof value === "string");
}

function looksLikeJson(value: string): boolean {
	const stripped = value.trimStart();
	return stripped.startsWith("{") || stripped.startsWith("[");
}

function codexEventName(point: HookInvocation["point"]): string {
	return ({
		pre_tool_use: "PreToolUse",
		post_tool_use: "PostToolUse",
		session_start: "SessionStart",
		user_prompt_submit: "UserPromptSubmit",
		stop: "Stop",
	} as Partial<Record<HookInvocation["point"], string>>)[point] ?? point;
}

function executionId(spec: ConfiguredHookSpec, invocation: HookInvocation): string {
	const canonical = [
		spec.scope,
		spec.hookId,
		spec.hookPoint,
		invocation.sessionId,
		invocation.turnId,
		invocation.toolName ?? "",
	].join("\0");
	return `hookexec_${createHash("sha256").update(canonical, "utf8").digest("hex").slice(0, 16)}`;
}

function isRecord(value: unknown): value is Readonly<Record<string, unknown>> {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}

function assertNotAborted(signal: AbortSignal): void {
	if (signal.aborted) throw abortError();
}

function abortError(): Error {
	const error = new Error("The operation was aborted");
	error.name = "AbortError";
	return error;
}
